File: test-diagnostics.py
#!/usr/bin/env python3
"""
Test script for diagnostics service
Fetches data from InfluxDB and runs diagnostics
"""

import requests
import json
from datetime import datetime, timedelta
from influxdb_client import InfluxDBClient

# Configuration
INFLUX_URL = "http://localhost:8086"
INFLUX_TOKEN = "o6cjAfkS_jFCvEePxDyz33zMQaJgbbSz_oqkSPzMTbROImhLlwDHwh8la4VMkMyNJsHWrVYs_JEHpWZGtFeaDw=="
INFLUX_ORG = "clpm"
INFLUX_BUCKET = "clpm_data"
DIAG_URL = "http://localhost:8050"

# Test loop
TEST_LOOP = "TIC208030"

def get_influx_data(loop_id, duration_minutes=15):
    """Fetch data from InfluxDB for the specified loop"""
    print(f"\n[DATA] Fetching data from InfluxDB for {loop_id}...")

    client = InfluxDBClient(url=INFLUX_URL, token=INFLUX_TOKEN, org=INFLUX_ORG)
    query_api = client.query_api()

    end_time = datetime.utcnow()
    start_time = end_time - timedelta(minutes=duration_minutes)

    query = f'''
    from(bucket: "{INFLUX_BUCKET}")
      |> range(start: {start_time.isoformat()}Z, stop: {end_time.isoformat()}Z)
      |> filter(fn: (r) => r._measurement == "control_loops")
      |> filter(fn: (r) => r.loop_id == "{loop_id}")
      |> filter(fn: (r) => r._field == "pv" or r._field == "op" or r._field == "sp")
      |> pivot(rowKey:["_time"], columnKey: ["_field"], valueColumn: "_value")
      |> filter(fn: (r) => exists r.pv and exists r.op)
      |> sort(columns: ["_time"])
    '''

    print(f"[TIME] Range: {start_time.isoformat()}Z to {end_time.isoformat()}Z")

    result = query_api.query(query)

    data = []
    for table in result:
        for record in table.records:
            data.append({
                'time': record.get_time().timestamp(),
                'pv': float(record.values.get('pv', 0)),
                'op': float(record.values.get('op', 0)),
                'sp': float(record.values.get('sp', 0)) if record.values.get('sp') else None
            })

    client.close()
    print(f"[OK] Retrieved {len(data)} data points from InfluxDB")
    return data

def run_diagnostics(loop_id, data):
    """Run diagnostics using the diagnostics service"""
    print(f"\n[DIAG] Running diagnostics for {loop_id}...")

    # Prepare series data
    series = {
        'ts': [d['time'] for d in data],
        'pv': [d['pv'] for d in data],
        'op': [d['op'] for d in data],
        'sp': [d['sp'] for d in data if d['sp'] is not None]
    }

    # Calculate sample rate
    sample_rate_hz = None
    if len(data) > 1:
        total_time = data[-1]['time'] - data[0]['time']
        sample_rate_hz = len(data) / total_time if total_time > 0 else None

    payload = {
        'loop_id': loop_id,
        'series': series,
        'sample_rate_hz': sample_rate_hz
    }

    print(f"[SUMMARY] Data summary:")
    print(f"   - Total points: {len(data)}")
    print(f"   - PV range: {min(series['pv']):.2f} to {max(series['pv']):.2f}")
    print(f"   - OP range: {min(series['op']):.2f} to {max(series['op']):.2f}")
    if sample_rate_hz:
        print(f"   - Sample rate: {sample_rate_hz:.4f} Hz")

    # Call diagnostics service
    response = requests.post(f"{DIAG_URL}/diagnostics/run", json=payload)
    response.raise_for_status()

    result = response.json()

    print(f"\n[RESULTS] Diagnostic Results:")
    print(f"   - Loop ID: {result['loop_id']}")
    print(f"   - Classification: {result['classification']}")
    print(f"   - Stiction (cross-correlation): {result['stiction_xcorr'] * 100:.2f}%")
    print(f"   - Oscillation Index: {result['osc_index']:.4f}")
    print(f"   - Oscillation Period: {result['osc_period_s']:.2f}s" if result['osc_period_s'] else "   - Oscillation Period: None detected")

    # Interpretation
    print(f"\n[INTERPRET] Interpretation:")
    if result['classification'] == 'stiction':
        print("   [WARNING] STICTION DETECTED - Valve may be sticking")
    elif result['classification'] == 'oscillating':
        print("   [WARNING] OSCILLATING - Loop is oscillating, may need tuning")
    elif result['classification'] == 'tuning':
        print("   [WARNING] TUNING ISSUE - Controller parameters may need adjustment")
    elif result['classification'] == 'deadband':
        print("   [WARNING] DEADBAND - Significant deadband detected")
    else:
        print("   [OK] NORMAL - Loop performance is acceptable")

    if result['stiction_xcorr'] > 0.35:
        print("   - High cross-correlation suggests valve stiction")
    if result['osc_index'] > 0.4:
        print("   - High oscillation index indicates sustained oscillations")

    return result

def main():
    print("[START] Starting Diagnostics Service Test\n")
    print("=" * 60)

    try:
        # Fetch data from InfluxDB
        influx_data = get_influx_data(TEST_LOOP, duration_minutes=15)

        if not influx_data:
            print("[ERROR] No data found in InfluxDB for the specified loop")
            return

        # Run diagnostics
        diagnostic_result = run_diagnostics(TEST_LOOP, influx_data)

        print("\n" + "=" * 60)
        print("[SUCCESS] Diagnostics test completed successfully!")
        print(f"\n[DB] Result that would be saved to database:")
        print(json.dumps(diagnostic_result, indent=2))

    except Exception as e:
        print(f"\n[ERROR] Test failed: {str(e)}")
        import traceback
        traceback.print_exc()
        exit(1)

if __name__ == "__main__":
    main()

File: test-diagnostics.js
const { InfluxDB } = require('@influxdata/influxdb-client');
const axios = require('axios');

// InfluxDB configuration
const INFLUX_URL = process.env.INFLUX_URL || 'http://localhost:8086';
const INFLUX_TOKEN = process.env.INFLUX_TOKEN || 'my-super-secret-auth-token';
const INFLUX_ORG = process.env.INFLUX_ORG || 'clpm';
const INFLUX_BUCKET = process.env.INFLUX_BUCKET || 'control_loops';

// Diagnostics service URL
const DIAG_URL = 'http://localhost:8050';

// Test loop configuration
const TEST_LOOP = {
  loop_id: 'TIC208030',
  name: 'TIC208030 Temperature Control Loop'
};

async function getInfluxData(loopId, durationMinutes = 15) {
  const influxDB = new InfluxDB({ url: INFLUX_URL, token: INFLUX_TOKEN });
  const queryApi = influxDB.getQueryApi(INFLUX_ORG);

  const endTime = new Date();
  const startTime = new Date(endTime.getTime() - durationMinutes * 60 * 1000);

  const fluxQuery = `
    from(bucket: "${INFLUX_BUCKET}")
      |> range(start: ${startTime.toISOString()}, stop: ${endTime.toISOString()})
      |> filter(fn: (r) => r.Tag == "${loopId}")
      |> filter(fn: (r) => r._field == "pv" or r._field == "op" or r._field == "sp" or r._field == "Mode")
      |> pivot(rowKey:["_time"], columnKey: ["_field"], valueColumn: "_value")
      |> filter(fn: (r) => exists r.pv and exists r.op)
  `;

  console.log(`\n📊 Fetching data from InfluxDB for ${loopId}...`);
  console.log(`⏰ Time range: ${startTime.toISOString()} to ${endTime.toISOString()}`);

  const data = [];

  return new Promise((resolve, reject) => {
    queryApi.queryRows(fluxQuery, {
      next(row, tableMeta) {
        const o = tableMeta.toObject(row);
        data.push({
          time: new Date(o._time).getTime() / 1000, // Unix epoch seconds
          pv: parseFloat(o.pv),
          op: parseFloat(o.op),
          sp: o.sp ? parseFloat(o.sp) : null,
          mode: o.Mode
        });
      },
      error(error) {
        console.error('❌ InfluxDB query error:', error);
        reject(error);
      },
      complete() {
        console.log(`✅ Retrieved ${data.length} data points from InfluxDB`);
        resolve(data);
      },
    });
  });
}

async function runDiagnostics(loopId, data) {
  console.log(`\n🔬 Running diagnostics for ${loopId}...`);

  // Prepare data in the format expected by diagnostics service
  const series = {
    ts: data.map(d => d.time),
    pv: data.map(d => d.pv),
    op: data.map(d => d.op),
    sp: data.filter(d => d.sp !== null).map(d => d.sp)
  };

  // Calculate sample rate
  const sampleRateHz = data.length > 1
    ? 1 / ((data[data.length - 1].time - data[0].time) / data.length)
    : null;

  const payload = {
    loop_id: loopId,
    series: series,
    sample_rate_hz: sampleRateHz
  };

  console.log(`📈 Data summary:`);
  console.log(`   - Total points: ${data.length}`);
  console.log(`   - PV range: ${Math.min(...series.pv).toFixed(2)} to ${Math.max(...series.pv).toFixed(2)}`);
  console.log(`   - OP range: ${Math.min(...series.op).toFixed(2)} to ${Math.max(...series.op).toFixed(2)}`);
  console.log(`   - Sample rate: ${sampleRateHz ? sampleRateHz.toFixed(4) : 'N/A'} Hz`);

  try {
    const response = await axios.post(`${DIAG_URL}/diagnostics/run`, payload);
    const result = response.data;

    console.log(`\n✨ Diagnostic Results:`);
    console.log(`   - Loop ID: ${result.loop_id}`);
    console.log(`   - Classification: ${result.classification}`);
    console.log(`   - Stiction (cross-correlation): ${(result.stiction_xcorr * 100).toFixed(2)}%`);
    console.log(`   - Oscillation Index: ${result.osc_index.toFixed(4)}`);
    console.log(`   - Oscillation Period: ${result.osc_period_s ? result.osc_period_s.toFixed(2) + 's' : 'None detected'}`);

    // Interpret results
    console.log(`\n🔍 Interpretation:`);
    if (result.classification === 'stiction') {
      console.log(`   ⚠️  STICTION DETECTED - Valve may be sticking`);
    } else if (result.classification === 'oscillating') {
      console.log(`   ⚠️  OSCILLATING - Loop is oscillating, may need tuning`);
    } else if (result.classification === 'tuning') {
      console.log(`   ⚠️  TUNING ISSUE - Controller parameters may need adjustment`);
    } else if (result.classification === 'deadband') {
      console.log(`   ⚠️  DEADBAND - Significant deadband detected`);
    } else {
      console.log(`   ✅ NORMAL - Loop performance is acceptable`);
    }

    if (result.stiction_xcorr > 0.35) {
      console.log(`   - High cross-correlation suggests valve stiction`);
    }
    if (result.osc_index > 0.4) {
      console.log(`   - High oscillation index indicates sustained oscillations`);
    }

    return result;
  } catch (error) {
    console.error('❌ Error calling diagnostics service:', error.response?.data || error.message);
    throw error;
  }
}

async function testDiagnostics() {
  console.log('🚀 Starting Diagnostics Service Test\n');
  console.log('=' .repeat(60));

  try {
    // Test 1: Get data from InfluxDB
    const influxData = await getInfluxData(TEST_LOOP.loop_id, 15);

    if (influxData.length === 0) {
      console.log('❌ No data found in InfluxDB for the specified loop');
      return;
    }

    // Test 2: Run diagnostics
    const diagnosticResult = await runDiagnostics(TEST_LOOP.loop_id, influxData);

    console.log('\n' + '='.repeat(60));
    console.log('✅ Diagnostics test completed successfully!');

  } catch (error) {
    console.error('\n❌ Test failed:', error.message);
    process.exit(1);
  }
}

// Run the test
testDiagnostics();
